src/lib/hooks/useBlockNumber.ts

```diff
@@ -0,0 +1,69 @@
+import useActiveWeb3React from 'hooks/useActiveWeb3React'
+import useDebounce from 'hooks/useDebounce'
+import useIsWindowVisible from 'hooks/useIsWindowVisible'
+import { atom } from 'jotai'
+import { useAtomValue, useUpdateAtom } from 'jotai/utils'
+import { useCallback, useEffect, useState } from 'react'
+
+function useBlock() {
+  const { chainId, library } = useActiveWeb3React()
+  const windowVisible = useIsWindowVisible()
+  const [state, setState] = useState<{ chainId?: number; block?: number }>({ chainId })
+
+  const onBlock = useCallback(
+    (block: number) => {
+      setState((state) => {
+        if (state.chainId === chainId) {
+          if (typeof state.block !== 'number') return { chainId, block }
+          return { chainId, block: Math.max(block, state.block) }
+        }
+        return state
+      })
+    },
+    [chainId]
+  )
+
+  useEffect(() => {
+    if (library && chainId && windowVisible) {
+      setState({ chainId })
+
+      library
+        .getBlockNumber()
+        .then(onBlock)
+        .catch((error) => {
+          console.error(`Failed to get block number for chainId ${chainId}`, error)
+        })
+
+      library.on('block', onBlock)
+      return () => {
+        library.removeListener('block', onBlock)
+      }
+    }
+    return undefined
+  }, [chainId, library, onBlock, windowVisible])
+
+  const debouncedBlock = useDebounce(state.block, 100)
+  return state.block ? debouncedBlock : undefined
+}
+
+const blockAtom = atom<number | undefined>(undefined)
+
+export function BlockUpdater() {
+  const setBlock = useUpdateAtom(blockAtom)
+  const block = useBlock()
+  useEffect(() => {
+    setBlock(block)
+  }, [block, setBlock])
+  return null
+}
+
+/** Requires that BlockUpdater be installed in the DOM tree. */
+export default function useBlockNumber(): number | undefined {
+  const { chainId } = useActiveWeb3React()
+  const block = useAtomValue(blockAtom)
+  return chainId ? block : undefined
```

Comment:
Will this behave strangely on other chains - shouldnt we index the block by chainId?

Reply:
If there's no chainId, or if the chainId is chainged, the atom should be cleared, so I think this will behave fine across chains. wdyt?

Reply:
Yep looks like in original state gets reset  in the useEffect when chainId changes